### ISSUE TITLE
Use Case: Expanding a matrix to fit incoming data

### ISSUE THREAD

**Issue by buddha314** [chapel-lang/chapel#7538]
I am ingesting the [Yummly data](https://www.kaggle.com/c/whats-cooking/data) as in #7524 .  This is a large array of JSON that includes thousands of recipes with their ingredients.  For instance, the first two recipes are 

```
{"recipes":
[
  {
    "id": 10259,
    "cuisine": "greek",
    "ingredients": [
      "romaine lettuce",
      "black olives",
      "grape tomatoes",
      "garlic",
      "pepper",
      "purple onion",
      "seasoning",
      "garbanzo beans",
      "feta cheese crumbles"
    ]
  },
  {
    "id": 25693,
    "cuisine": "southern_us",
    "ingredients": [
      "plain flour",
      "ground pepper",
      "salt",
      "tomatoes",
      "ground black pepper",
      "thyme",
      "eggs",
      "green tomatoes",
      "yellow corn meal",
      "milk",
      "vegetable oil"
    ]
  }]}
```

My intent is to turn this into a Graph via a sparse matrix.  The matrix should be square, the size of the number of unique ingredients in the recipes per side.  The entries of the sparse matrix should be the number of times "yellow corn meal" co-occurs with "vegetable oil" or null.

Also, I need to be able to retrieve the corresponding row/name.  I need to know that "vegetable oil" is row and column 19, or know that 19 corresponds to "vegetable oil".

I do not know the number of ingredients or even the number of recipes in advance.

Here Be Dragons.  I don't know if this is the right approach, so maybe ignore this bit.

First, I thought I'd create a domain `D = {1..0, 1..0}` and `SD: sparse subdomain(D)`. I did a loop through recipes and then through ingredients.  When a new ingredient was found, I tried doing `D = {1..D.shape[0]+1, 1..D.shape[0]+1}`.  Later, I wanted the cooccurence, so I tried

```
for in1 in recipe.ingredients {
  for in2 in recipe.ingredients {
    SD += (ingredients[in1], ingredients[in1]);
  }
}
```

but immediately got 

```
yummlyAnalysis.chpl:53: error: halt reached - Sparse domain/array index out of bounds: (1, 1) (expected to be within {1..0, 1..0})
```

It feels just out of my grasp.

**Comment by mppf** [comment]
I'm not sure why resizing D didn't work, but I would personally probably try to write it to compute the maximum ingredient number first, then construct D, then construct the sparse subdomain. Of course there may be some application reason not to do that.

**Comment by buddha314** [comment]
These datasets can get massive, it seems inefficient to have to pass over it twice.  I notice that most of the examples I see on SO assume you know the size of the matrix in advance, and that prevents you from dynamically loading data.  Once the data is in, that's a different story.

**Comment by mppf** [comment]
Could you arbitrarily use D = {1..max(int), 1..max(int)} ?

**Comment by buddha314** [comment]
I think that is similar to @bradcray 's suggestion.  I don't really know, except that it seems pretty awkward.  How does Python do it under the covers?

**Comment by buddha314** [comment]
Also, if you do it that way, couldn't we just instantiate that domain for a vanilla sparse subdomain()?

**Comment by bradcray** [comment]
On IRC yesterday, @benharsh noted that DefaultSparse does track its bounding box dynamically but that CS* layouts don't.  I don't believe there's anything inherent that would prevent them from doing so, simply that the engineering effort hasn't been put in.  Needless to say, like any array reallocation, it'll cost something in performance for the productivity of having things grow beautifully (though possibly not much).

To @buddha314's question about creating a sparse subdomain over a {1..bigint, 1..bigint} bounding box, the main downside is that it would require O(bigint) storage even if you only used {1..11, 1..11} of the bounding box.  You'd also be less likely to get OOB errors if the programmer made a mistake.

Bottom line: I think we can make CS* layouts track the parent bounding box and probably should.

**Comment by buddha314** [comment]
Well, cool!  So it wasn't my imagination, it worked on regular domains.  Let me try that now because my current data probably does NOT need to be distributed.

**Comment by bradcray** [comment]
Note that it's not a question of distribution, but rather of the local memory layout.  IIRC, you've been using CS* because it's what LinearAlgebra wants, not because of anything related to distribution.

Your note about distribution is an interesting one, though, because the block distribution doesn't (currently at least) support the ability to change the distribution dynamically.  So this tracking of the bounding box would definitely raise load balancing challenges in the distributed case, at least for Block.

**Comment by buddha314** [comment]
Yes, my abuse of the language.  This particular matrix should not be too big for even my laptop.  The next one I expect will be.  I'm starting with the Yummly data to try out a method.

If you have a good example of the "best" way to do this, I'd love to see it.  It looks like just removing `use LayoutCS` has worked, but it's not done yet.

**Comment by bradcray** [comment]
Yeah, for a shared-memory code, I would expect it to "just work" (at least, based on what @benharsh said).

To see the challenge with a distributed matrix, consider the following example:  Say you start with a 20x20 sparse matrix on 4 nodes:  You'd like each node to own an equal amount, say 10x10 or 20x5 or 5x20.  So you could get this by setting up a block distribution with a 20x20 bounding box.  But now say you want to grow it to a 22x22 matrix.  With a block distribution, the 20x20 bounding box implies that the new rows/columns will all be owned by the same nodes, so some nodes will own a few more rows/cols than others.  2 more rows/cols isn't a big deal, but once you're at 100x100 or 1000x1000 you've got a serious load imbalance.  One way to fix this _might_ be to have the block distribution support changing its bounding box, but I'm not sure this is really desirable because the amount of fine-grain data shuffling that would need to take place between the nodes.

So, another way to go would be to use a Cyclic or Block-Cyclic distribution, which do a better job of load balancing as things grow without shuffling data around because of their round-robin nature.  At present, though, neither of these distributions supports sparse matrices (but could... it's just work that hasn't been taken on yet).

**Comment by buddha314** [comment]
Would it make sense to provide something for matrices like `A.finalize()` that you could run and have it redistribute before you try calculations against it?

**Comment by buddha314** [comment]
Does this method make sense for the time being?

```
  var ingredients: domain(string),
      ingredientIds: [ingredients] int,
      ings = {1..0},
      D = {ings.dim(1),ings.dim(1)},
      //SD: sparse subdomain(D) dmapped CS(),
      SD: sparse subdomain(D),
      A: [SD] real;

...

      for recipe in cookBook.recipes {
        //writeln("\t", recipe.cuisine);
        for ingredient in recipe.ingredients {
          if !ingredients.member(ingredient) {
            ings = {1..ings.last + 1};
            writeln("D is now ", D);
            ingredients.add(ingredient);
            ingredientIds[ingredient] = ingredients.size;
          }
        }
        for ingredient_1 in recipe.ingredients {
          for ingredient_2 in recipe.ingredients {
            //writeln(D);
            var i = ingredientIds(ingredient_1);
            var j = ingredientIds(ingredient_2);
            writeln("Trying to add to SD: (i,j) ", i, " ", j);
            SD += (ingredientIds(ingredient_1), ingredientIds(ingredient_2));
            writeln("SD shape: ", SD.shape);
          }
        }
      }
```

I've been looking at this too long.  The `writeln` for D is not showing an increment.

**Comment by benharsh** [comment]
@buddha314 : In your latest code, it looks like you're trying to modify ``ings`` in order to resize ``D``. When you wrote
```chpl
D = {ings.dim(1), ings.dim(1)};
```
You used the range *values* of ``ings``. In other words, there's no link between ``ings`` and ``D`` and I don't think there ever will be.

@bradcray : Looks like I was mistaken. It seems DefaultSparse simply doesn't do as much error checking as it should, because the actual ``parentDom`` it stores internally is not changed. I assumed it changed because code like this worked:
```chpl
var parent = {1..0, 1..0};
var D : sparse subdomain(parent);

// This program completes successfully even if this
// line is commented out...
parent = {1..100, 1..100};
D += (99, 99);
```
It seems that we don't bounds check the indices in DefaultSparseDom.dsiAdd.

**Comment by bradcray** [comment]
> Would it make sense to provide something for matrices like A.finalize() that you could run and have it redistribute before you try calculations against it?

I don't think this would help with the block-distributed case because the problem is less about being unable to redistribute at the end, and more about not being sure where to store things while you're building things up.  If you could afford to store sufficient state in local memory prior to distributing it and declaring arrays over it, then I think a similar approach would work in which you'd read in the data, determine the problem size, declare the distributed domains, populate them, and then declare and populate the arrays.  But whether or not that's scalable depends a lot on how much data you need to have in memory in order to create those distributed declarations.

**Comment by benharsh** [comment]
A potential alternative to linking the original parent domain could be to utilize the ``domain.expand`` method, which if used today would result in a compile-time error:

```
expand not supported on sparse domains
```

``domain.expand`` on sparse domains could only modify the bounds.

**Comment by bradcray** [comment]
> Looks like I was mistaken.

I opened issue #7544 to capture the question that I think is at the core here:  Sparse domains don't dynamically track their parent domains—should they?  I also wedged a concept in there about a sparse domain that didn't have a notion of a parent domain (which might be more what Brian wants when he refers to a more Python-like experience).

**Comment by bradcray** [comment]
(BTW, @buddha314, let me know if the new issue captures the feature request here well enough to replace this issue).

**Comment by buddha314** [comment]
@benharsh Thanks. I need to find (or perhaps someone can create) a canonical example for people adding data to domains of un-predetermined size.  Which sounds very [Ed and Lorraine Warren](https://en.wikipedia.org/wiki/Ed_and_Lorraine_Warren) I know, but is a common example.  When loading data, you don't know how much you're going to get.

@bradcray I think you made it more specific than I was thinking, but that's good since you'll have to do the translation.  I don't feel too justified in offering an opinion beyond the use case.

**Comment by ben-albrecht** [comment]
> I opened issue #7544 to capture the question that I think is at the core here

Going to close this as duplicate with the more specific issue #7544.

Let me know if you disagree.
